### PR TITLE
Occupancy Grid Maps

### PR DESCRIPTION
**Pull request by judicaelclair**
- 2D & 3D Occupancy Grid Maps with visualisation, sample maps
- Robust `SimulatorConfig`, properly handles outdated config (e.g. map / algorithm names no longer exist).
- Invisible colour import error fix.

Credit goes to Abel & John for implementing the `OccupancyGridMap` class.

### CODE REVIEW

File: src/structures/dynamic_colour.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 from typing import Optional, Callable
 
-from structures.colour import Colour
+from structures.colour import Colour, TRANSPARENT
```
Comment (JohnYao93) [inline src/structures/dynamic_colour.py:3]: why do you need this

File: src/simulator/views/map/map_view.py [hunk 2]
```diff
@@ -225,7 +241,7 @@ def eager_refresh():
             for p in np.ndindex(self.map.traversables_data.shape):
                 self.__cube_modified[p] = self.map.traversables_data[p]
         
-        clr = self._services.state.effective_view.colours[MapData.TRAVERSABLES]()
+        clr = TRANSPARENT if self.__is_ogm else self._services.state.effective_view.colours[MapData.TRAVERSABLES]()
```
Comment (JohnYao93) [inline src/simulator/views/map/map_view.py:244]: are we no longer able to change the color of the traversables in an ogm?

File: src/simulator/views/gui/simulator_config.py [hunk 3]
```diff
@@ -395,23 +397,32 @@ def __init__(self, services: Services, mouse1_press_callbacks: List[Callable[[],
...
         for so in self.__services.state.objects:
             if isinstance(so, SimulatorConfigState):
                 self.__state = so
                 cmd = self.__maps_option['command']
-                self.__maps_option.set(self.__map_keys.index(so.mp))
-                self.__maps_option['command'] = cmd
-                self.__algorithms_option.set(self.__algorithm_keys.index(so.algo))
-                self.__animations_option.set(self.__animation_keys.index(so.ani))
-                self.__update_position_entries()
+                try:
```
Comment (Nabushika) [inline src/simulator/views/gui/simulator_config.py:405]: Do we need this? Seems a little hacky... If its necessary, can we have a specific exception to catch, thanks

**Review by Nabushika — APPROVED** [review]
One minor change

File: src/simulator/views/map/map_view.py (hunk 2 above)
Comment (JohnYao93) [reply src/simulator/views/map/map_view.py:244]: ie this looks like the color of traversables is just hard set to transparent? maybe im wrong though

File: src/structures/dynamic_colour.py (hunk 1 above)
Comment (judicaelclair) [reply src/structures/dynamic_colour.py:3]: import error

File: src/simulator/views/map/map_view.py (hunk 2 above)
Comment (judicaelclair) [reply src/simulator/views/map/map_view.py:244]: Fair point actually, we should probably still be able to see the base colour `traversables` and people can then toggle that to invisible if they like right?

File: src/simulator/views/gui/simulator_config.py (hunk 3 above)
Comment (judicaelclair) [reply src/simulator/views/gui/simulator_config.py:405]: Seriously what is your problem with exceptions based on completely unknown, this is totally fine.

File: src/simulator/views/map/map_view.py (hunk 2 above)
Comment (JohnYao93) [reply src/simulator/views/map/map_view.py:244]: yeah i dont think it should be a special case; only the obstacles should change in an ogm imo. perhaps we can set the default of all the other 3d map views of traversables to transparent as well btw?

Comment (JohnYao93) [reply src/simulator/views/map/map_view.py:244]: cuz i agree its hard to see the obstacles in an ogm if the traversables arent transparent; but its also the case for the 3d maps as well

File: src/simulator/views/gui/simulator_config.py (hunk 3 above)
Comment (judicaelclair) [reply src/simulator/views/gui/simulator_config.py:407]: Changed it to output the exception in terminal.

File: src/simulator/views/map/map_view.py (hunk 2 above)
Comment (judicaelclair) [reply src/simulator/views/map/map_view.py]: Changed to have no special treatment.

**Review by JohnYao93 — APPROVED** [review]
lgtm

**Pull request merged**